File: heap_perm_gen.lua
--Permutation generator in Lua
--Heap's algorithm
--Modified from [1]
--References:
--[1] http://ww.lua.org/pil/9.3.html
--[2] https://en.wikipedia.org/wiki/Heap's_algorithm



function permgen (a, n)
  if n == 1 then
    coroutine.yield(a)
  else
    for i=1,n-1 do
      permgen(a,n-1)

      --if n is even then swap(A[i], A[n])
      --else swap(A[1], A[n])
      j = (n%2 == 0 and i or 1)
      a[j],a[n] = a[n], a[j]
      
    end
    permgen(a, n-1)
  end
end

function perm (a)
  local n = rawlen(a)
  return coroutine.wrap(function () permgen(a, n) end)
end

function printResult (a)
  for i,v in ipairs(a) do
    io.write(v, " ")
  end
  io.write("\n")
end

--print permutations of abcd
for p in perm{"a", "b", "c", "d"} do
  printResult(p)
end

File: heap_perm_gen.js
// Permutation generator in ES6
// References:
// https://en.wikipedia.org/wiki/Heap's_algorithm

'use strict'
function * heapPerm (a, n) {
  if (n === 1) yield a
  else {
    for (let i = 1, j; i <= n; i++) {
      yield * heapPerm(a, n - 1)
      if (n % 2 === 1) j = 1
      else j = i
      let swap = a[j - 1]
      a[j - 1] = a[n - 1]
      a[n - 1] = swap
    }
  }
}

// test
var a = [1, 2, 3, 4]
for (let ele of heapPerm(a, a.length)) {
  console.log(ele)
}
